main.go

```diff
@@ -43,26 +46,123 @@ func main() {
+		registry.MustRegister(client)
+		http.Handle(*basicMetricsPathF, promhttp.HandlerFor(registry, promhttp.HandlerOpts{
 			ErrorLog:      log.NewErrorLogger(),
 			ErrorHandling: promhttp.ContinueOnError,
 		}))
 	}
 
+	// This collector should be only one for both cases, because it creates goroutines which sends API requests to Amazon in background.
+	enhancedCollector := enhanced.NewCollector(sess)
+
 	// enhanced metrics
+	// NOTE: This handler was retained for backward compatibility. See: https://jira.percona.com/browse/PMM-1901.
 	{
 		registry := prometheus.NewRegistry()
-		registry.MustRegister(enhanced.NewCollector(sess))
+		registry.MustRegister(enhancedCollector)
 		http.Handle(*enhancedMetricsPathF, promhttp.HandlerFor(registry, promhttp.HandlerOpts{
 			ErrorLog:      log.NewErrorLogger(),
 			ErrorHandling: promhttp.ContinueOnError,
 		}))
 	}
 
-	log.Infof("Basic metrics   : http://%s%s", *listenAddressF, *basicMetricsPathF)
-	log.Infof("Enhanced metrics: http://%s%s", *listenAddressF, *enhancedMetricsPathF)
+	// all metrics (with filtering)
+	{
+		f := factory.New(cfg, sess, map[string]prometheus.Collector{"enhanced": enhancedCollector, "client": client})
+		handler := newHandler(f)
+		http.Handle("/metrics", handler)
+	}
+
+	log.Infof("Metrics: http://%s%s", *listenAddressF, "/metrics")
 	log.Fatal(http.ListenAndServe(*listenAddressF, nil))
 }
+
+// handler wraps an unfiltered http.Handler but uses a filtered handler,
+// created on the fly, if filtering is requested. Create instances with
+// newHandler. It used for collectors filtering.
+type handler struct {
+	unfilteredHandler http.Handler
+	factory           *factory.Collectors
+}
+
+func newHandler(factory *factory.Collectors) *handler {
+	h := &handler{factory: factory}
+	if innerHandler, err := h.innerHandler(); err != nil {
+		log.Fatalf("Couldn't create metrics handler: %s", err)
+	} else {
+		h.unfilteredHandler = innerHandler
+	}
+	return h
+}
+
+// ServeHTTP implements http.Handler.
+func (h *handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	filters := r.URL.Query()["collect[]"]
+	log.Debugln("collect query:", filters)
+
+	if len(filters) == 0 {
+		// No filters, use the prepared unfiltered handler.
+		h.unfilteredHandler.ServeHTTP(w, r)
+		return
+	}
+
+	filteredHandler, err := h.innerHandler(filters...)
+	if err != nil {
+		log.Warnln("Couldn't create filtered metrics handler:", err)
+		w.WriteHeader(http.StatusBadRequest)
+		_, err := w.Write([]byte(fmt.Sprintf("Couldn't create filtered metrics handler: %s", err)))
+		if err != nil {
+			log.Errorln(err)
+		}
+		return
+	}
+	filteredHandler.ServeHTTP(w, r)
+}
+
+func (h *handler) innerHandler(filters ...string) (http.Handler, error) {
+	registry := prometheus.NewRegistry()
+	registry.MustRegister(prometheus.NewProcessCollector(os.Getpid(), "")) // from prometheus.DefaultGatherer
+	registry.MustRegister(prometheus.NewGoCollector())                     // from prometheus.DefaultGatherer
```

Comment:
`standard.process`, `standard.go` as in postgres_exporter. `?collect[]=badname` should return zero metrics.